lib/model/query/entities.js

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 ODK Central Developers
+// See the NOTICE file at the top-level directory of this distribution and at
+// https://github.com/getodk/central-backend/blob/master/NOTICE.
+// This file is part of ODK Central. It is subject to the license terms in
+// the LICENSE file found in the top-level directory of this distribution and at
+// https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+// including this file, may be copied, modified, propagated, or distributed
+// except according to the terms contained in the LICENSE file.
+
+const { sql } = require('slonik');
+const { Entity } = require('../frames');
+const { getEntity } = require('../../data/entity');
+const { unjoiner } = require('../../util/db');
+
+
+const _defInsert = (id, partial, json) => sql`insert into entity_defs ("entityId", "submissionDefId", "data", "current", "createdAt")
+  values (${id}, ${partial.def.submissionDefId}, ${json}, true, clock_timestamp())
+  returning *`;
+
+const nextval = sql`nextval(pg_get_serial_sequence('entities', 'id'))`;
+
+const _getDataset = (dsName, subDefId) => sql`
+SELECT datasets."id", sd."id" as "subDefId" FROM
+submission_defs AS sd
+  JOIN form_defs AS fd ON sd."formDefId" = fd."id"
+  JOIN forms AS f ON fd."formId" = f."id"
+  JOIN datasets ON datasets."projectId" = f."projectId"
+WHERE datasets."name" = ${dsName} AND sd."id" = ${subDefId}`;
+
+const createNew = (partial) => ({ one }) => {
+  const json = JSON.stringify(partial.def.data);
+
+  return one(sql`
+with def as (${_defInsert(nextval, partial, json)}),
+ds as (${_getDataset(partial.aux.dataset, partial.def.submissionDefId)}),
+ins as (insert into entities (id, "datasetId", "uuid", "label", "createdAt")
```

Comment:
I'm going to leave this as is because it's inserting a mix of values from different sources (a previous query, the partial entity, the functions arguments).